src/ray/rpc/raylet/raylet_client.cc

```diff
@@ -0,0 +1,400 @@
+    for (size_t i = 0; i < reply.fractional_resource_ids().size(); ++i) {
+      auto const &fractional_resource_ids = reply.fractional_resource_ids()[i];
+      auto &acquired_resources = resource_ids_[fractional_resource_ids.resource_name()];
+
+      // Each resource includes a series of resource IDs (e.g., GPU 0) and corresponding
+      // amount for that resource ID. If the resource amount is fractional, then there
+      // should only be one resource ID.
+      size_t num_resource_ids = fractional_resource_ids.resource_ids().size();
+      size_t num_resource_fractions = fractional_resource_ids.resource_fractions().size();
+      RAY_CHECK(num_resource_ids == num_resource_fractions);
+      RAY_CHECK(num_resource_ids > 0);
+      for (size_t j = 0; j < num_resource_ids; ++j) {
+        int64_t resource_id = fractional_resource_ids.resource_ids()[j];
+        double resource_fraction = fractional_resource_ids.resource_fractions()[j];
+        if (num_resource_ids > 1) {
+          int64_t whole_fraction = resource_fraction;
+          RAY_CHECK(whole_fraction == resource_fraction);
+        }
+        acquired_resources.emplace_back(resource_id, resource_fraction);
+      }
+    }
+    task_spec->reset(new ray::TaskSpecification(reply.task_spec()));
+  } else {
+    *task_spec = nullptr;
+    RAY_LOG(INFO) << "Failed to get task, msg: " << status.error_message();
+  }
+  return GrpcStatusToRayStatus(status);
+}
+
+ray::Status RayletClient::TaskDone() {
+  RETURN_IF_DISCONNECTED(is_connected_);
+  TaskDoneRequest task_done_request;
+  task_done_request.set_worker_id(worker_id_.Binary());
+
+  auto callback = [](const Status &status, const TaskDoneReply &reply) {
+    if (!status.ok()) {
+      RAY_LOG(INFO) << "Failed to send TaskDoneRequest, msg: " << status.message();
+    }
+  };
+
+  auto call =
+      client_call_manager_.CreateCall<RayletService, TaskDoneRequest, TaskDoneReply>(
+          *stub_, &RayletService::Stub::PrepareAsyncTaskDone, task_done_request,
+          callback);
+  return call->GetStatus();
+}
+
+ray::Status RayletClient::FetchOrReconstruct(const std::vector<ObjectID> &object_ids,
+                                             bool fetch_only,
+                                             const TaskID &current_task_id) {
+  RETURN_IF_DISCONNECTED(is_connected_);
+  FetchOrReconstructRequest fetch_or_reconstruct_request;
+  fetch_or_reconstruct_request.set_fetch_only(fetch_only);
+  fetch_or_reconstruct_request.set_task_id(current_task_id.Binary());
+  fetch_or_reconstruct_request.set_worker_id(worker_id_.Binary());
+  IdVectorToProtobuf<ObjectID, FetchOrReconstructRequest>(
+      object_ids, fetch_or_reconstruct_request,
+      &FetchOrReconstructRequest::add_object_ids);
+
+  // Callback to deal with reply.
+  auto callback = [this](const Status &status, const FetchOrReconstructReply &reply) {
+    if (!status.ok() && is_connected_) {
+      is_connected_ = false;
```

Comment:
For other async requests, we should also check `call->GetStatus()` and set `is_connected_` to false if not ok.